cli/NetworkIO.hpp

```diff
@@ -0,0 +1,227 @@
+};
+}  // namespace networkio_internal
+
+class NetworkCliServiceImpl final : public NetworkCli::Service {
+ public:
+  NetworkCliServiceImpl()
+    : NetworkCli::Service(),
+      worker_exclusive_mtx_(),
+      running_(true),
+      request_state_() { }
+
+  /**
+   * Handles gRPC request. Sets the buffer in the RequestState, notifies the main thread, then waits for a response.
+   * @param context
+   * @param request
+   * @param response
+   * @return
+   */
+  Status SendQuery(grpc::ServerContext *context,
+                   const QueryRequest *request,
+                   QueryResponse *response) override {
+    auto service_ex_lock = enter();
+    if (!service_ex_lock) {
+      return Status::CANCELLED;
+    }
+
+    // Service thread sets a notification for the main thread.
+    // Because of service thread exclusivity, we have rights to this data structure.
+    request_state_.requestReady(request->query());
+
+    // Service thread - main thread critical section:
+    // Service thread waits for the buffered message response from the main thread. The main thread will set
+    // consumer_ready_ when it is finished and released its exclusive lock on the communication data structure.
+    std::unique_lock<std::mutex> lock(request_state_.mutex_);
+    while (!request_state_.response_ready_)
+      request_state_.condition_.wait(lock);
+
+    *response = request_state_.response_message_;
+
+    return Status::OK;
+  }
+
+  void kill() {
+    running_ = false;
+  }
+
+  networkio_internal::RequestState& getRequestState() {
+    return request_state_;
+  }
+
+ private:
+  /**
+   * When a worker enters, it gains exclusive access to the main thread. That is, no other worker of this service
+   * is allowed to interact with main thread.
+   * @return A lock which grants the worker mutual exclusion.
+   */
+  std::unique_lock<std::mutex> enter() {
+    std::unique_lock<std::mutex> lock(worker_exclusive_mtx_);
+    if (!running_)
+      lock.unlock();
+    return lock;
+  }
+
+  std::mutex worker_exclusive_mtx_;
+  bool running_;
+  networkio_internal::RequestState request_state_;
+};
+
+class NetworkIO : public IOInterface {
+ public:
+  NetworkIO()
+      : IOInterface(),
+        out_stream_(),
+        err_stream_(),
+        server_(nullptr),
+        service_() {
+    std::string server_address("0.0.0.0:" + std::to_string(FLAGS_port));
+
+    // Starts a server.
+    grpc::ServerBuilder builder;
+    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
+    builder.RegisterService(&service_);
+    server_ = builder.BuildAndStart();
+    LOG(INFO) << "Listening on port " << std::to_string(FLAGS_port);
+  }
+
+  FILE* out() override {
+    return out_stream_.file();
+  }
+
+  FILE* err() override {
+    return err_stream_.file();
+  }
+
+  std::string getNextCommand() override {
+    // critical section: wait for a command
+    networkio_internal::RequestState& request_state = service_.getRequestState();
```

Comment:
Change to `networkio_internal::RequestState &request_state`.

Ditto in `notifyCommandComplete`.